shared_model/backend/protobuf/queries/impl/proto_query.cpp

```diff
@@ -4,92 +4,140 @@
+                     shared_model::proto::GetAccountTransactions,
+                     shared_model::proto::GetAccountAssetTransactions,
+                     shared_model::proto::GetTransactions,
+                     shared_model::proto::GetAccountAssets,
+                     shared_model::proto::GetAccountDetail,
+                     shared_model::proto::GetRoles,
+                     shared_model::proto::GetRolePermissions,
+                     shared_model::proto::GetAssetInfo,
+                     shared_model::proto::GetPendingTransactions>;
+
+  /// list of types in proto variant
+  using ProtoQueryListType = ProtoQueryVariantType::types;
+}  // namespace
 
 namespace shared_model {
   namespace proto {
 
-    template <typename QueryType>
-    Query::Query(QueryType &&query)
-        : CopyableProto(std::forward<QueryType>(query)),
-          variant_{[this] {
-            auto &&ar = *proto_;
-            int which = ar.payload()
-                            .GetDescriptor()
-                            ->FindFieldByNumber(ar.payload().query_case())
-                            ->index_in_oneof();
-            return shared_model::detail::
-                variant_impl<Query::ProtoQueryListType>::template load<
-                    Query::ProtoQueryVariantType>(
-                    std::forward<decltype(ar)>(ar), which);
-          }},
-          ivariant_{detail::makeLazyInitializer(
-              [this] { return QueryVariantType(*Query::variant_); })},
-          blob_{[this] { return makeBlob(*proto_); }},
-          payload_{[this] { return makeBlob(proto_->payload()); }},
-          signatures_{[this] {
-            SignatureSetType<proto::Signature> set;
-            if (proto_->has_signature()) {
-              set.emplace(proto_->signature());
-            }
-            return set;
-          }} {}
-
-    template Query::Query(Query::TransportType &);
-    template Query::Query(const Query::TransportType &);
-    template Query::Query(Query::TransportType &&);
-
-    Query::Query(const Query &o) : Query(o.proto_) {}
-
-    Query::Query(Query &&o) noexcept : Query(std::move(o.proto_)) {}
+    struct Query::Impl {
+      explicit Impl(TransportType &&ref) : proto_{std::move(ref)} {}
+      explicit Impl(const TransportType &ref) : proto_{ref} {}
+
+      detail::ReferenceHolder<TransportType> proto_;
+
+      ProtoQueryVariantType variant_{[this] {
+        auto &&ar = *proto_;
+        int which = ar.payload()
+                        .GetDescriptor()
+                        ->FindFieldByNumber(ar.payload().query_case())
+                        ->index_in_oneof();
+        return shared_model::detail::variant_impl<ProtoQueryListType>::
+            template load<ProtoQueryVariantType>(std::forward<decltype(ar)>(ar),
+                                                 which);
+      }()};
+
+      QueryVariantType ivariant_{
+          [this] { return QueryVariantType(variant_); }()};
```

Comment:
I understand you did not add this code in this PR, but the lambda seems useless here, if we can just write
`QueryVariantType ivariant_{QueryVariantType(variant_)};`

Reply:
Maybe even simpler in this case
`QueryVariantType ivariant_{variant_};`

Reply:
yes)